numpy/core/shape_base.py

```diff
@@ -443,9 +452,131 @@ def _atleast_nd(a, ndim):
     return array(a, ndmin=ndim, copy=False, subok=True)
 
 
+def _accumulate(values):
+    # Helper function because Python 2.7 doesn't have
+    # itertools.accumulate
+    value = 0
+    accumulated = []
+    for v in values:
+        value += v
+        accumulated.append(value)
+    return accumulated
+
+
+def _concatenate_shapes(shapes, axis):
+    """Given array shapes, return the resulting shape and slices prefixes.
+
+    These help in nested concatation.
+    Returns
+    -------
+    shape: tuple of int
+        This tuple satisfies:
+        ```
+        shape, _ = _concatenate_shapes([arr.shape for shape in arrs], axis)
+        shape == concatenate(arrs, axis).shape
+        ```
+
+    slice_prefixes: tuple of (slice(start, end), )
+        For a list of arrays being concatenated, this returns the slice
+        in the larger array at axis that needs to be sliced into. 
+
+        For example, the following holds:
+        ```
+        ret = concatenate([a, b, c], axis)
+        _, (sl_a, sl_b, sl_c) = concatenate_slices([a, b, c], axis)
+        
+        ret[(slice(None),) * axis + sl_a] == a
+        ret[(slice(None),) * axis + sl_b] == b
+        ret[(slice(None),) * axis + sl_c] == c
+        ```
+        
+        Thses are called slice prefixes since they are used in the recursive
+        blocking algorithm to compute the left most slice during the
+        recursion. Therefore, they must be prepended to rest of the slice
+        that was computed deeper in the recusion.
+        
+        These are returned as tuples to ensure that they can quickly be added
+        to existing slice tuple without creating a new tuple everytime.
+
+    """
+    # Cache a result that will be reused.
+    shape_at_axis = [shape[axis] for shape in shapes]
+
+    # Take a shape, any shape
+    first_shape = shapes[0]
+    first_shape_pre = first_shape[:axis]
+
+    # We don't need to check ``shape[axis+1:] == first_shaoe[axis+1:]``
+    # as it is be checked in the inner parts of the recursion
```

Comment:
I just noticed this comment. I think you need to check those axes too:

```python
import numpy as np
from numpy.core.shape_base import _block_setup, _block_slicing
ones = np.ones

def blk(arrays):
    arrays, list_ndim, result_ndim, final_size = _block_setup(arrays)
    return _block_slicing(arrays, list_ndim, result_ndim)

print(blk([[ones((2,3)), ones((2,2))], [ones((2,2)), ones((2,2))]]))
```
output:
```
[[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 0.]
 [1. 1. 1. 1. 0.]]
```

Reply:
I added it as a test. Good one.